drivers/video/video_ctrls.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/video.h>
+#include <zephyr/drivers/video-controls.h>
+#include <zephyr/logging/log.h>
+
+#include "video_async.h"
+#include "video_ctrls.h"
+#include "video_device.h"
+
+LOG_MODULE_REGISTER(video_ctrls, CONFIG_VIDEO_LOG_LEVEL);
+
+void video_ctrl_handler_init(struct video_ctrl_handler *hdl, const struct device *dev)
+{
+	if (!hdl) {
+		return;
+	}
+
+	hdl->dev = dev;
+
+	sys_slist_init(&hdl->list);
+}
+
+void video_ctrl_handler_free(struct video_ctrl_handler *hdl)
+{
+	if (!hdl) {
+		return;
+	}
+
+	hdl->dev = NULL;
+
+	while (sys_slist_get(&hdl->list)) {
+	}
+}
+
+void video_ctrl_handler_add(struct video_ctrl_handler *hdl, struct video_ctrl_handler *added)
+{
+	if (!hdl || !added || hdl == added || sys_slist_is_empty(&added->list)) {
+		return;
+	}
+
+	sys_slist_append_list(&hdl->list, sys_slist_peek_head(&added->list),
+			      sys_slist_peek_tail(&added->list));
+}
+
+void video_ctrl_handler_remove(struct video_ctrl_handler *hdl, struct video_ctrl_handler *removed)
+{
+	sys_snode_t *node;
+
+	if (!hdl || !removed || sys_slist_is_empty(&hdl->list) ||
+	    sys_slist_is_empty(&removed->list)) {
+		return;
+	}
+
+	SYS_SLIST_FOR_EACH_NODE(&removed->list, node) {
+		sys_slist_find_and_remove(&hdl->list, node);
+	}
+}
+
+static inline int check_range(enum video_ctrl_type type, int32_t min, int32_t max, uint32_t step,
+			      uint32_t def)
+{
+	switch (type) {
+	case VIDEO_CTRL_TYPE_BOOLEAN:
+		if (step != 1 || max > 1 || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_INTEGER:
+	case VIDEO_CTRL_TYPE_INTEGER64:
+		if (step == 0 || min > max || def < min || def > max) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_MENU:
+	case VIDEO_CTRL_TYPE_INTEGER_MENU:
+		if (min > max || def < min || def > max || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	default:
+		return 0;
+	}
+}
+
+static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32_t *flags)
+{
+	*flags = 0;
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+	case VIDEO_CID_VFLIP:
+		*type = VIDEO_CTRL_TYPE_BOOLEAN;
+		break;
+	case VIDEO_CID_POWER_LINE_FREQUENCY:
+	case VIDEO_CID_TEST_PATTERN:
+		*type = VIDEO_CTRL_TYPE_MENU;
+		break;
+	case VIDEO_CID_PIXEL_RATE:
+		*type = VIDEO_CTRL_TYPE_INTEGER64;
+		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+		break;
+	default:
+		*type = VIDEO_CTRL_TYPE_INTEGER;
+		break;
```

Comment:
I like the fact that controls are integer by default. This means only exceptions have to be maintained, and `VIDEO_CTRL_TYPE_INTEGER` fits most things.